### PR TITLE
KEYCLOAK-13962  SAML2 Identity Provider - During login phase, SamlAuthenticationPreprocessors are not taken into account to produce an appropriate destination url

### PR DESCRIPTION
**Pull request by looorent**
## Problem

During a SAML login, the `destinationUrl` used to post the SAML request is not affected by the `org.keycloak.protocol.saml.preprocessor.SamlAuthenticationPreprocessor`.

If a SAML preprocessor modifies a destination url, the request should be posted accordingly.

## Solution

When posting the SAML request, use the destination url available on the request itself and not a variable defined earlier.

## Steps to reproduce

* Define a SamlAuthenticationPreprocessor that modifies the request's destinationUrl (for instance, by adding a query parameter to the url.
* When the request is posted, this new url will not be used

### CODE REVIEW

**Comment by hmlnarik** [comment]
Thank you for the PR. As a start, it looks good.

If the `destinationUrl` is `null`, the flow needs to be cancelled (see e.g. [here](https://github.com/keycloak/keycloak/blob/10.0.0/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L161))

The same situation is for other SAML message types (logout request, logout response) that can be modified via a preprocessor, and applies also to standard clients as well as identity providers. Could you please update the other places as well?

**Comment by looorent** [comment]
Thanks for you review. I can adapt this.

To cancel a flow, your example suggests breaking the current method execution using `return`. It this the way to go in Keycloak or should I throw a specific exception instead?

**Comment by hmlnarik** [comment]
Thank you. Best is to follow the same flow as is used in the respective method. Some of the methods throw an exception (e.g. `SAMLIdentityProvider.performLogin(AuthenticationRequest`), some just return (e.g. `SAMLIdentityProvider.backchannelLogout`).
If some case would be uncertain, feel free to submit the code with a comment in the appropriate place, and we'll work it out case by case.

**Comment by looorent** [comment]
I have implemented a post-check for null destinations (everywhere a pre-processor is involved).
However, several tests do not pass, for example this (self-explanatory) one: https://github.com/keycloak/keycloak/blob/f28ca30e6df92f85fb9037729c710d5728dffed5/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/BasicSamlTest.java#L122-L133

It fails because I have added this piece of code ( [there]( https://github.com/keycloak/keycloak/blob/32f13016fafcaadec9092a8b078d7a7717a18b29/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L377) ):
```
 if (requestAbstractType.getDestination() == null) {
                event.detail(Details.REASON, "no_destination_url_in_login_request");
                event.error(Errors.INVALID_SAML_AUTHN_REQUEST);
                return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.INVALID_REQUEST);
}
```

Are you sure the destination is mandatory in every situation? This test has a different assumption.

**Comment by hmlnarik** [comment]
> I have implemented a post-check for null destinations (everywhere a pre-processor is involved).
> However, several tests do not pass, for example this (self-explanatory) one:
> 
> https://github.com/keycloak/keycloak/blob/f28ca30e6df92f85fb9037729c710d5728dffed5/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/BasicSamlTest.java#L122-L133
> 
> It fails because I have added this piece of code ( [there](https://github.com/keycloak/keycloak/blob/32f13016fafcaadec9092a8b078d7a7717a18b29/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L377) ):
> 
> ```
>  if (requestAbstractType.getDestination() == null) {
>                 event.detail(Details.REASON, "no_destination_url_in_login_request");
>                 event.error(Errors.INVALID_SAML_AUTHN_REQUEST);
>                 return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.INVALID_REQUEST);
> }
> ```
> 
> Are you sure the destination is mandatory in every situation? This test has a different assumption.

No, some of the SAML message types have destination optional. The check needs to be in line with SAML spec, see https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf, lines 661-664 and 843-846, and a related PR #7022

**Comment by looorent** [comment]
Thanks for your feedback, @hmlnarik.
The code have been adapted and all tests are now ok.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 1]
```diff
@@ -121,6 +125,10 @@ public Response performLogin(AuthenticationRequest request) {
                 authnRequest = it.next().beforeSendingLoginRequest(authnRequest, request.getAuthenticationSession());
             }
 
+            destinationUrl = ofNullable(authnRequest.getDestination())
+                    .map(URI::toString)
+                    .orElseThrow(() -> new IdentityBrokerException("No destination in login request"));
+
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:131]: There should be no exception thrown. The destination is attribute only mandatory in certain cases, see the SAML spec above.
The code should be similar to:
```java
if (authnRequest.getDestination() != null) {
    destinationUrl = authnRequest.getDestination().toString();
}
```

Similar code needs to be in `keycloakInitiatedBrowserLogout` and `backchannelLogout` methods.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 2]
```diff
@@ -212,6 +220,9 @@ protected LogoutRequestType buildLogoutRequest(UserSessionModel userSession, Uri
         for (Iterator<SamlAuthenticationPreprocessor> it = SamlSessionUtils.getSamlAuthenticationPreprocessorIterator(session); it.hasNext();) {
             logoutRequest = it.next().beforeSendingLogoutRequest(logoutRequest, userSession, null);
         }
+        if (logoutRequest.getDestination() == null) {
+            throw new IdentityBrokerException("No destination in logout request");
+        }
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:225]: Remove, destination is optional per spec.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java [hunk 3]
```diff
@@ -538,6 +538,10 @@ public ResponseType transformLoginResponse(List<ProtocolMapperProcessor<SAMLLogi
             response = (ResponseType) it.next().beforeSendingResponse(response, clientSessionCtx.getClientSession());
         }
 
+        if (response.getDestination() == null) {
+            throw new RuntimeException("No destination in login response");
+        }
+
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:544]: Remove, destination is optional per spec.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java [hunk 4]
```diff
@@ -745,6 +749,9 @@ protected LogoutRequestType createLogoutRequest(String logoutUrl, AuthenticatedC
         for (Iterator<SamlAuthenticationPreprocessor> it = SamlSessionUtils.getSamlAuthenticationPreprocessorIterator(session); it.hasNext();) {
             logoutRequest = it.next().beforeSendingLogoutRequest(logoutRequest, clientSession.getUserSession(), clientSession);
         }
+        if (logoutRequest.getDestination() == null) {
+            throw new RuntimeException("No destination in logout request");
+        }
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:754]: Remove, destination is optional per spec.

File: services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java [hunk 5]
```diff
@@ -523,6 +523,9 @@ public Response authenticated(BrokeredIdentityContext context) {
             for(Iterator<SamlAuthenticationPreprocessor> it = SamlSessionUtils.getSamlAuthenticationPreprocessorIterator(session); it.hasNext();) {
                 loginResponse = it.next().beforeProcessingLoginResponse(loginResponse, authenticationSession);
             }
+            if (loginResponse.getDestination() == null) {
+                throw new RuntimeException("No destination in login response");
+            }
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java:528]: Remove, destination is optional per spec.

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 6]
```diff
@@ -334,6 +334,12 @@ protected Response logoutRequest(LogoutRequestType request, String relayState) {
                 }
             }
 
+            if (request.getDestination() == null) {
+                event.detail(Details.REASON, "no_destination_in_logout_request");
+                event.error(Errors.INVALID_SAML_AUTHN_REQUEST);
+                return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.INVALID_REQUEST);
+            }
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java:341]: Remove, destination is optional per spec.

**Review by hmlnarik — CHANGES_REQUESTED** [review]

**Comment by looorent** [comment]
Thanks @hmlnarik.
Everything has been updated accordingly.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 7]
```diff
@@ -162,6 +167,12 @@ public void backchannelLogout(KeycloakSession session, UserSessionModel userSess
         JaxrsSAML2BindingBuilder binding = buildLogoutBinding(session, userSession, realm);
         try {
             LogoutRequestType logoutRequest = buildLogoutRequest(userSession, uriInfo, realm, singleLogoutServiceUrl);
+            if (logoutRequest.getDestination() != null) {
+                singleLogoutServiceUrl = logoutRequest.getDestination().toString();
+            } else {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:172]: Please do remove the `else` part. In case destination is not set, the original value of `singleLogoutServiceUrl` should be used for submission.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 8]
```diff
@@ -186,6 +197,11 @@ public Response keycloakInitiatedBrowserLogout(KeycloakSession session, UserSess
        } else {
             try {
                 LogoutRequestType logoutRequest = buildLogoutRequest(userSession, uriInfo, realm, singleLogoutServiceUrl);
+                if (logoutRequest.getDestination() != null) {
+                    singleLogoutServiceUrl = logoutRequest.getDestination().toString();
+                } else {
+                    return null;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:203]: Same as above

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 9]
```diff
@@ -54,6 +54,7 @@
 import java.util.Set;
 import java.util.TreeSet;
 
+
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:57]: Remove unrelated change

**Review by hmlnarik — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java (hunk 7 above)
Comment (looorent) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:172]: I managed this case exactly as it is done a few lines above. This method does not support blank values for `singleLogoutServiceUrl`, should we allow blank values after building the logout request anyway?

https://github.com/keycloak/keycloak/blob/f8a4f66d6c183de894a8a436f9e35dfa3248e862/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L163

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java (hunk 8 above)
Comment (looorent) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:203]: Same answer as above. This is inspired from this existing line:

https://github.com/keycloak/keycloak/blob/f8a4f66d6c183de894a8a436f9e35dfa3248e862/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L183

**Review by looorent — COMMENTED** [review]
Thanks for your comments. I have inspired these lines from existing null checks. Should we allow blank values anyway?

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java (hunk 7 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:172]: By this line, the `singleLogoutServiceUrl` is guaranteed to be non-null and non-empty exactly because of the mentioned check. Hence use the value from the `singleLogoutServiceUrl`.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java [hunk 10]
```diff
@@ -745,7 +745,7 @@ protected LogoutRequestType createLogoutRequest(String logoutUrl, AuthenticatedC
         for (Iterator<SamlAuthenticationPreprocessor> it = SamlSessionUtils.getSamlAuthenticationPreprocessorIterator(session); it.hasNext();) {
             logoutRequest = it.next().beforeSendingLogoutRequest(logoutRequest, clientSession.getUserSession(), clientSession);
         }
-        
+
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:748]: Remove unrelated change

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java (hunk 7 above)
Comment (looorent) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:172]: Based on [this comment](https://github.com/keycloak/keycloak/pull/7007#discussion_r452036172), I understood that you wanted me to put the request's destination in`singleLogoutServiceUrl` because custom processors could be involved in `buildLogoutRequest`  (as we do for the destination url in the login flow). I can remove this, no problem.

**Comment by looorent** [comment]
I have updated everything accordingly.
Let me know whether I can rebase master and squash everything.

**Review by hmlnarik — APPROVED** [review]
Thank you for the contribution!
Travis build: https://travis-ci.com/github/keycloak/keycloak/builds/175685950
